### ISSUE TITLE
How to choose margin and scale parameters correctly in arcface loss?

### ISSUE THREAD

**Issue by RocketFlash** [KevinMusgrave/pytorch-metric-learning#186]
Hello @KevinMusgrave ! 
Arcface loss has margin and scale parameters, could you give some advices how to choose them correctly?

**Comment by KevinMusgrave** [comment]
Let E<sub>A</sub> be the embeddings for class A, and W<sub>A</sub> be the learned classifier weights for class A.

Here's how ArcFace compares with normalized cross entropy (i.e. cross entropy loss with normalized weights and embeddings):

**Normalized cross entropy loss**: 
 - The angle between E<sub>A</sub> and W<sub>A</sub> should be smaller than the angle between E<sub>A</sub> and W<sub>B</sub>, E<sub>A</sub> and W<sub>C</sub> ...

**ArcFace**: 
 - The angle between E<sub>A</sub> and W<sub>A</sub> should be smaller than the angle between E<sub>A</sub> and W<sub>B</sub>, E<sub>A</sub> and W<sub>C</sub> ... **by the specified margin**.

So increasing the margin results in a bigger separation of classes in your training set. I don't think there's an easy way to choose the value, because it depends on how many classes you're training on, how different/similar the classes are, and how much you need your embeddings to generalize. 

The scale parameter changes the shape of the logits. The higher the scale, the more peaky the logits vector becomes. A higher scale parameter will result in larger gradients during training. Note that "softmax scale" is the inverse of "softmax temperature". Again I don't think there is an obvious way of setting this parameter. The ArcFace paper uses 64 because CosFace used 64, and they used 64 through trial and error.

**Comment by RocketFlash** [comment]
@KevinMusgrave Thank you for detailed answer!

I found a paper [AdaCos: Adaptively Scaling Cosine Logits for Effectively Learning Deep Face Representations](https://arxiv.org/pdf/1905.00292.pdf), where authors provide some advices on choosing **m** and **s** parameters. They suggest choosing a parameter **s** like sqrt(2)*(log(C-1)), where C - number of classes. It is not clear why in the cosine base metric articles the scale parameter is chosen equal to 64.

In original ArcFace paper authors refer to [CosFace: Large Margin Cosine Loss for Deep Face Recognition](https://arxiv.org/pdf/1801.09414.pdf) and justify the choice of parameters based on this article. The problem is that based on this article, the **m** parameter should be <<1 (much less than 1) if number of classes is large, but they choose m=0.5.

**Comment by KevinMusgrave** [comment]
Interesting choice for **s**. Re: ArcFace and CosFace, does the parameter range provided in the CosFace paper transfer to ArcFace though? The ArcFace margin is an angle inside the cosine expression, whereas the CosFace margin is a cosine value.

**Comment by RocketFlash** [comment]
Yes, I know, this makes me even more confused. From ArcFace paper:

"We follow [CosFace: Large Margin Cosine Loss for Deep Face Recognition] to set the feature scale s to 64 and choose
the angular margin m of ArcFace at 0.5."

**Comment by KevinMusgrave** [comment]
I think they're just talking about **s** when they say "we follow CosFace..."
